rfcs/0008-program-testing-strategy/README.md

```diff
@@ -0,0 +1,88 @@
+
+**Problems**: Without systematic testing, we lack confidence in initial releases and subsequent changes. Regressions in program eligibility go undetected until production. Manual testing effort scales linearly with program count, becoming unsustainable.
+
+**Motivation**: Implementing a comprehensive testing strategy now—before further scaling—will compound benefits: faster development cycles, reduced production bugs, and easier refactoring. This RFC establishes testing as a first-class requirement for program development.
+
+## Proposal
+
+Implement a testing strategy that emphasizes unit testing (Unit > Integration > E2E) and with specific standards for backend and frontend code. Use fixture-based testing to isolate external API dependencies (PolicyEngine) and enable fast, reliable test execution. Use code coverage reports to evaluate progress.
+
+## Implementation
+
+### Backend Testing
+
+**Unit Tests**: Test program eligibility and value calculation logic in isolation.
+
+**Integration Tests**: Test API endpoints, multi-model interactions, and service layer operations. Use `vcrpy` fixtures for PolicyEngine calls. Verify correct request formation and response handling.
+
+### Frontend Testing
+
+**Unit Tests**: 
+There isn't much (if any) FE code added per program, but there are a few core behaviors that we should add tests for as a one-off task:
+- Test pure business logic functions for income calculations (`calcIncomeStreamAmount`, `calcMemberYearlyIncome`, `calcTotalIncome`) and value formatting (`programValue`, `calculateTotalValue`).
+
+**Integration Tests**: Test english translations, program configuration, and "current benefit" flows. Use Playwright with HAR recording for backend API fixtures to enable offline testing.
+
+### E2E Testing
+
+Use existing patterns to write a limited number smoke tests per white label configuration. Tests run against dev/staging databases without fixtures to catch integration issues. Implement a periodic cleanup job to remove test records.
+
+### Code Coverage
+
+Code coverage reports are not a substitute for a team culture that values testing. They are, however, a tool we can use to gauge progress, raise concerns, and guide conversation.
+
+**Tools**:
+- Backend: `coverage.py` for Python coverage reporting
+- Frontend: Istanbul/nyc (existing)
+- CI: CodeCov (free for open source) for tracking trends and (optionally) README badges
+
+## Rollout Plan
+
+### Phase 1
+- Ensure all tests run in CI
+    * Add GH workflow to run BE test suite
+    * Add GH workflow to run FE unit tests
+    * Fix/Comment out existing test failures
+- Add coverage tooling
+    * Add `coverage.py` for BE coverage reporting
+    * Add GH workflow to upload BE coverage report to CodeCov
+    * Add GH workflow to upload FE coverage report to CodeCov
```

Comment:
I could see an argument for moving this to **Phase 2** if we wanted to get to the new architecture and TX implementation ASAP.

I have it in **Phase 1** because we want our TX implementation to be exemplary and I think it'll help us ensure our tests are comprehensive.